src/bin/pgcopydb/ld_stream.c

```diff
@@ -767,8 +775,49 @@ streamRotateFile(LogicalStreamContext *context)
 		return true;
 	}
 
-	/* compute the WAL filename that would host the current LSN */
-	XLByteToSeg(context->cur_record_lsn, segno, context->WalSegSz);
+	/*
+     * determine the LSN to calculate walFileName in which to write the current
+     * message.
+     *
+	 * This walFileName calculation later ensures safe transaction formation in
+	 * the transform/apply process by always appending messages here to the
+	 * latest file and preventing rotation to earlier files.
+	 *
+	 * In most cases, jsonFileLSN should be the same as cur_record_lsn. However,
+	 * occasionally, current messages may have LSNs lower than the previous
+	 * ones. This can occur due to concurrent transactions with interleaved
+	 * LSNs. Since the logical decoding protocol sends the complete transaction
+	 * at commit time, the LSNs for messages within one transaction may be lower
+	 * than those of the previous transactions. In such cases, we use maximum
```

Comment:
I think it's more helpful to say: the LSNs for messages within one transaction may be lower than the COMMIT lsn of the previously streamed transaction. At least for me it makes things more obvious/immediate. See what you feel about it.